File: PAT_Basic_Level_by_lua/1096 大美数/1096 大美数.lua
-- 题目编号: 1096 大美数
-- 实现原理: 枚举N的四个不同正因子，判断N是否能被它们的和整除。
local k = tonumber(io.read())  -- 读取待检测的数字个数k

-- 对每个数字进行检测
for i = 1, k do
    local n = io.read("*n")  -- 当前待检测的数N
    local a = {}  -- 存储N的全部因子
    
    -- 从1开始枚举，找N的因子
    for d = 1, n do
        if n % d == 0 then
            a[#a + 1] = d
        end
    end
    
    local ok = false
    for p = 1, #a - 3 do
        for q = p + 1, #a - 2 do
            for r = q + 1, #a - 1 do
                for u = r + 1, #a do
                    if n % (a[p] + a[q] + a[r] + a[u]) == 0 then
                        ok = true
                    end
                end
            end
        end
    end
    print(ok and "Yes" or "No")
end

File: PAT_Basic_Level_by_ruby/1096 大美数/1096 大美数.rb
# 1096 大美数
#
# 题目描述：
#   若正整数 N 可以整除它的 4 个不同正因数之和，则称这样的正整数为“大美数”。本题就要求你判断任一给定的正整数是否是“大美数”。
#
# 实现原理：
#   找出所有能整除该数的因子；判断是否存在4个因子之和等于该数本身；枚举因子组合验证条件
#

if __FILE__ == $PROGRAM_NAME
  data = STDIN.read.split.map(&:to_i)
  k = data.shift
  data.first(k).each do |n|
    divisors = []
    (1..Math.sqrt(n).to_i).each do |d|
      next unless (n % d).zero?

      divisors << d
      divisors << n / d if d * d != n
    end
    puts(divisors.combination(4).any? { |four| four.sum % n == 0 } ? 'Yes' : 'No')
  end
end
